messages/main.go

```diff
@@ -22,51 +24,82 @@ type Response events.APIGatewayProxyResponse
 
 // Handler is our lambda handler invoked by the `lambda.Start` function call
 func Handler(request Request) (Response, error) {
-	buf := new(bytes.Buffer)
 	body := []byte(request.Body)
+	log.Printf("Receiving request body %s", body)
 	resp := Response{
 		IsBase64Encoded: false,
 		Headers: map[string]string{
 			"Content-Type": "application/json",
 		},
 	}
-	eventsAPIEvent, e := slackevents.ParseEvent(json.RawMessage(body),
-		slackevents.OptionVerifyToken(&slackevents.TokenComparator{VerificationToken: slackSigningSecret}))
+	eventsAPIEvent, e := slackevents.ParseEvent(
+		json.RawMessage(body),
+		slackevents.OptionVerifyToken(&slackevents.TokenComparator{VerificationToken: slackSigningSecret}),
+	)
 	if e != nil {
+		log.Println("Could not parse Slack event :'(")
 		resp.StatusCode = 500
 	}
+	log.Printf("Processing an event of outer type %s", eventsAPIEvent.Type)
 
-	if eventsAPIEvent.Type == slackevents.Message {
-		var r *slackevents.ChallengeResponse
-		err := json.Unmarshal(body, &r)
-		if err != nil {
-			resp.StatusCode = 500
-		}
-		resp.Headers["Content-Type"] = "text"
-		buf.Write([]byte(r.Challenge))
-		resp.Body = buf.String()
-	}
-	if eventsAPIEvent.Type == slackevents.CallbackEvent {
+	handleSlackChallenge(eventsAPIEvent, body, resp)
+
+	if eventsAPIEvent.Type == slackevents.CallbackEvent || eventsAPIEvent.Type == slackevents.AppMention {
 		innerEvent := eventsAPIEvent.InnerEvent
+		log.Printf("Processing an event of inner data %s", innerEvent.Data)
 		switch ev := innerEvent.Data.(type) {
 		case *slackevents.AppMentionEvent:
-			api.PostMessage(ev.Channel, slack.MsgOptionText("Yes, hello.", false))
+			resp.StatusCode = 200
+			fmt.Printf("Reacting to app mention event from channel %s", ev.Channel)
+			_, _, postError := botApi.PostMessage(ev.Channel, slack.MsgOptionText("Yes, hello.", false))
+			if postError != nil {
+				resp.StatusCode = 500
+				log.Printf("Error while posting message %s", postError)
+			}
 		case *slackevents.AppHomeOpenedEvent:
-			api.PublishView(
+			fmt.Println("Reacting to app home request event")
+			resp.StatusCode = 200
+			_, publishViewError := botApi.PublishView(
 				ev.User,
-				slack.HomeTabViewRequest{"home", apphome.UserHome("Cyril").Blocks, "", "", ""},
-				ev.View.Hash)
+				slack.HomeTabViewRequest{
+					Type:   "home",
+					Blocks: apphome.UserHome("Cyril").Blocks,
+				},
+				ev.View.Hash,
+			)
+			if publishViewError != nil {
+				resp.StatusCode = 500
+				log.Printf("Error while posting message %s", publishViewError)
+			}
 		}
 	}
-
 	return resp, nil
+}
 
+// Slack Challenge is used to register the URL in the slack API config interface
+// Should only be used once by slack when changing the events URL
+func handleSlackChallenge(eventsAPIEvent slackevents.EventsAPIEvent, body []byte, resp Response) {
+	if eventsAPIEvent.Type == slackevents.URLVerification {
+		buf := new(bytes.Buffer)
+		var r *slackevents.ChallengeResponse
+		err := json.Unmarshal(body, &r)
```

Comment:
This will fail because r is null at this point

Reply:
why is that ? Json.unmarshal unmarshals the content of the first variable in the second one, so it's good ?
(this code actually ran to validate the endpoint on the slack API website, so unless I messed something while refactoring, it should still work)

Reply:
Oh yes, you're right mb